### ISSUE TITLE
Add support for groups of custom fields

### ISSUE THREAD

**Issue by jedelman8** [nautobot/nautobot#899]
### Overview

Often times when there is a large number of CFs, it becomes unwieldy to manage and view them in the UI. Users often have some namespace in the CF name that in reality _is a group_.  It would be nice to create an optional group when creating new CFs similar to custom links.  From a UI perspective, we can create a table per group or a column that shows the group.

~Going one step further, if we seperate the step to create the group, this would allow users to apply a group of custom fields to more than one model more easily.  If there is a "role" or other unique "ID" and you need to apply those two CFs to 5 models, that would be an iterative process in the UI. If we went down this path, users would create a group of 2 CFs: role and ID and then apply that group to any model they'd like.~

What do others think?  Is there a need to ~create groups of CFs to easily apply them to more than one model~, improve the UI, both, or neither? =) 

Updating the issue with strike through text after re-scoping - see comments below.

**Comment by glennmatthews** [comment]
What kind of quantity of distinct custom fields on a specific model have we encountered in real-world deployments? A dozen? A hundred? Just trying to understand the magnitude of the problem that we're looking to solve here.

**Comment by itdependsnetworks** [comment]
I would think this would apply to every single implementation, simply looking at only two points, with devices and roles. Roles inherently require different information, take a simple matrix of the following.

|   Role | BGP ASN |  STP Root | Multicast ID |
| ------------- | ------------- | ------------- | ------------- |
| Backbone Router  | Y  | N  | Y |
| Internet Router | Y  | N  | N |
| Campus agg switch  | N  | Y  | N |
| Campus l2 switch | N  | Y | N |

**Comment by glennmatthews** [comment]
> I would think this would apply to every single implementation, simply looking at only two points, with devices and roles. Roles inherently require different information...

Sorry for being dense, but I don't quite follow how this matrix applies to this issue. You're describing 3 custom fields, right? Which apply differently to one of four different device roles? I don't see any grouping of fields as applicable or needed here - was this comment meant to apply to #889 instead?

**Comment by itdependsnetworks** [comment]
Sorry, you are right, that comment should be for the feature in #889 . To this one, I have not seen this as an issue yet within environments I have worked in.

**Comment by jedelman8** [comment]
> What kind of quantity of distinct custom fields on a specific model have we encountered in real-world deployments? A dozen? A hundred? Just trying to understand the magnitude of the problem that we're looking to solve here.

Hard to estimate that, but I hope others will chime in.

**Comment by brobare** [comment]
I believe that I would have a use-case for this. One idea I had was to use custom fields to store "metadata" on the device model that would have some significance in external tooling. This metadata would be used by that tool to set some logic or behavior based on those metadata values. 

For example, I may want to use this metadata to describe monitoring profiles to the NMS. These monitoring profiles would have say, a packet loss threshold set to them. I may create a field and call it "device_metric_profile_loss" and make it an integer. In the NMS platform, it would read this CF and use it to decide when to threshold for packet loss. If I set it to 2%, the NMS wont alert for threshold crossings until >2%. Another NMS related CF may be "device_metric_profile_latency" and make it an integer. If I set the latency to 600, the NMS wont alert until the latency crosses 600ms. Another might be a "device_down_threshold" field for alerting. Set it to 15 and the NMS does not alert until a device is down for 15 minutes. All of these would have some significance within the NMS platform, and could be grouped, from a UI perspective, together. Perhaps an accordion type card?

Another independent group of data on the device model may be escalation information for a device. Create a CF that I may call "cf_escalation_hours", make it a text field, and have notes about when (or when not) to escalate. Create "cf_escalation_contact" and provide contact info for that device. Create a "cf_support_contact" and use that to store information about the NOC/support contact for a device.


Other examples I could see this being useful for is the Tenant model. Create custom fields to describe information about that tenant. Perhaps best illustrated by a picture:
<img width="961" alt="Screen Shot 2021-09-10 at 10 24 15 PM" src="https://user-images.githubusercontent.com/53182374/132937430-a25d7160-59f9-4eb7-b409-5ebe83f26047.png">

This is already a (somewhat) shortened list as adding all of our customer data led to long pages. Several of these could be grouped together and put into a UI element that allowed for collapsing.

> What kind of quantity of distinct custom fields on a specific model have we encountered in real-world deployments? A dozen? A hundred? Just trying to understand the magnitude of the problem that we're looking to solve here.

For me, I could see it being a half a dozen groups of half a dozen entries per group (per model). So, not hundreds, but, a few dozen.

Does this line up with what you were thinking, @jedelman8?

**Comment by jedelman8** [comment]
Yes, this is great @brobare .  From the write up, I can't tell which is more important for your use case.  Is it more important to you to improve the UI with groups of CFs (Escalation, NMS, etc.) vs. being able to apply that group of CFs more easily to different models.  Could you do without the latter or is that still a nice to have?

**Comment by brobare** [comment]
@jedelman8 Yes, my thought was just the ability to group CFs on a model, then have those groups rendered in a UI element along the lines of:
https://bbbootstrap.com/snippets/accordion-menu-dropdown-font-awesome-icons-93684856
If I have 6 groups of 6, without something like this, I'd have 36 rows in the CF panel in the UI. With some sort of collapsable, only 6 rows would display (each of the group names) on load, then clicking on the group would expand it, showing the 6 children for that group. I'd just need a way to tell each of the 36 CF's which group they belong to. Similar to the custom links behavior.

For your "latter" scenario, I hadnt really thought about that. Was the idea there to have CF groups defined in a way that could be abstract from the models, then applied to one or more model? Seems like if you had CF groups that always had the same list of CF elements, could save a lot of work in building it out.

If I understand the latter correctly, that'd be a nice to have and could do without it. 

Could definitely see it being useful though. Especially if CF groups could be defined as JSON, could have a lot of common CF scenarios templatized, then imported and applied as needed. For example:
```
{
    "cf_group_name": "cf_contacts",
    "cf_group_label": "Contacts"
    "custom_fields": [
        {
            "type": "url",
            "name": "website",
            "weight": 100,
            "label": "Website",
            "description": "",
            "filter_logic": "Loose"
        },
        {   
            "type": "text",
            "name": "email",
            "weight": 100,
            "label": "Contact email",
            "description": "",
            "filter_logic": "Loose"
        },
        {   
            "type": "text",
            "name": "phone_number",
            "weight": 100,
            "label": "Phone",
            "description": "",
            "filter_logic": "Loose"
        },
        {   
            "type": "boolean",
            "name": "after_hours",
            "weight": 100,
            "label": "After hours support?",
            "description": "",
            "filter_logic": "Loose"
        },
        {   
            "type": "choices",
            "name": "Type",
            "weight": 100,
            "label": "Customer type",
            "description": "",
            "filter_logic": "Loose",
            "choices": [
                {
                    "value": "Enterprise",
                    "weight": 100
                },
                {
                    "value": "Wholesale",
                    "weight": 101
                },
                {
                    "value": "Education",
                    "weight": 102
                }
            ]
        }
    ]
}
```
I think could be made generic enough to apply to multiple models, then applied as needed. 

Then if it could be built in a way that you can apply multiple of these JSON defined CF groups, the CF panel would be really quick to be built in a new install. And even faster, if there was a library of these CF group JSON objects defined.

Maybe this makes sense :)

**Comment by jathanism** [comment]
Assertions based on the comment thread to scope this:

- Add a field to custom field objects to allow an optional "group" to be defined
- Augment the custom field panel in the UI detail view for objects to dynamically render fields using nested tables, by their group if it is set

**Comment by bryanculver** [comment]
Would it be required to know what group a custom field key would be in just by looking at the object itself?

API only returns:
```
"custom_fields": {
    "emails_circuit_maintenances": null,
    "provider_parser_circuit_maintenances": null
},
```

**Comment by lampwins** [comment]
I think all custom fields should always be included in the API. I could see a versioned API here that groups custom fields in the API by they group name:

```
"custom_fields": {
    "default": {
        "i_dont_have_a_group": true
    },
    "group_a": {    
        "emails_circuit_maintenances": null,
        "provider_parser_circuit_maintenances": null
    }
},
```

**Comment by glennmatthews** [comment]
If we're going to go in that direction, then we need to think about group "label" versus group "slug", and I'm not sure the added complexity is really worthwhile.

**Comment by bryanculver** [comment]
Closed via #2317
